Fix all warnings from -Wcast-align

mattkelly opened the pull request:
This fixes a very large amount of annoying warnings generated by -Wcast-align. Building with clang-207.0.31 on OS X.

Example of warnings that are fixed in ccm_memory.c:

```
src/encauth/ccm/ccm_memory.c:259:23: warning: cast from 'unsigned char *' to 'LTC_FAST_TYPE *' (aka 'unsigned long long *') increases required
      alignment from 1 to 8 [-Wcast-align]
                    *((LTC_FAST_TYPE*)(&PAD[z]))  ^= *((LTC_FAST_TYPE*)(&pt[y+z]));
                      ^~~~~~~~~~~~~~~~~~~~~~~~~
src/encauth/ccm/ccm_memory.c:259:56: warning: cast from 'unsigned char *' to 'LTC_FAST_TYPE *' (aka 'unsigned long long *') increases required
      alignment from 1 to 8 [-Wcast-align]
                    *((LTC_FAST_TYPE*)(&PAD[z]))  ^= *((LTC_FAST_TYPE*)(&pt[y+z]));
                                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~
src/encauth/ccm/ccm_memory.c:260:23: warning: cast from 'unsigned char *' to 'LTC_FAST_TYPE *' (aka 'unsigned long long *') increases required
      alignment from 1 to 8 [-Wcast-align]
                    *((LTC_FAST_TYPE*)(&ct[y+z])) = *((LTC_FAST_TYPE*)(&pt[y+z])) ^ *((LTC_FAST_TYPE*)(&CTRPAD[z]));
                      ^~~~~~~~~~~~~~~~~~~~~~~~~~
src/encauth/ccm/ccm_memory.c:260:55: warning: cast from 'unsigned char *' to 'LTC_FAST_TYPE *' (aka 'unsigned long long *') increases required
      alignment from 1 to 8 [-Wcast-align]
                    *((LTC_FAST_TYPE*)(&ct[y+z])) = *((LTC_FAST_TYPE*)(&pt[y+z])) ^ *((LTC_FAST_TYPE*)(&CTRPAD[z]));
                                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~
src/encauth/ccm/ccm_memory.c:260:87: warning: cast from 'unsigned char *' to 'LTC_FAST_TYPE *' (aka 'unsigned long long *') increases required
      alignment from 1 to 8 [-Wcast-align]
                    *((LTC_FAST_TYPE*)(&ct[y+z])) = *((LTC_FAST_TYPE*)(&pt[y+z])) ^ *((LTC_FAST_TYPE*)(&CTRPAD[z]));
                                                                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/encauth/ccm/ccm_memory.c:279:23: warning: cast from 'unsigned char *' to 'LTC_FAST_TYPE *' (aka 'unsigned long long *') increases required
      alignment from 1 to 8 [-Wcast-align]
                    *((LTC_FAST_TYPE*)(&pt[y+z])) = *((LTC_FAST_TYPE*)(&ct[y+z])) ^ *((LTC_FAST_TYPE*)(&CTRPAD[z]));
                      ^~~~~~~~~~~~~~~~~~~~~~~~~~
src/encauth/ccm/ccm_memory.c:279:55: warning: cast from 'unsigned char *' to 'LTC_FAST_TYPE *' (aka 'unsigned long long *') increases required
      alignment from 1 to 8 [-Wcast-align]
                    *((LTC_FAST_TYPE*)(&pt[y+z])) = *((LTC_FAST_TYPE*)(&ct[y+z])) ^ *((LTC_FAST_TYPE*)(&CTRPAD[z]));
                                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~
src/encauth/ccm/ccm_memory.c:279:87: warning: cast from 'unsigned char *' to 'LTC_FAST_TYPE *' (aka 'unsigned long long *') increases required
      alignment from 1 to 8 [-Wcast-align]
                    *((LTC_FAST_TYPE*)(&pt[y+z])) = *((LTC_FAST_TYPE*)(&ct[y+z])) ^ *((LTC_FAST_TYPE*)(&CTRPAD[z]));
                                                                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/encauth/ccm/ccm_memory.c:280:23: warning: cast from 'unsigned char *' to 'LTC_FAST_TYPE *' (aka 'unsigned long long *') increases required
      alignment from 1 to 8 [-Wcast-align]
                    *((LTC_FAST_TYPE*)(&PAD[z]))  ^= *((LTC_FAST_TYPE*)(&pt[y+z]));
                      ^~~~~~~~~~~~~~~~~~~~~~~~~
src/encauth/ccm/ccm_memory.c:280:56: warning: cast from 'unsigned char *' to 'LTC_FAST_TYPE *' (aka 'unsigned long long *') increases required
      alignment from 1 to 8 [-Wcast-align]
                    *((LTC_FAST_TYPE*)(&PAD[z]))  ^= *((LTC_FAST_TYPE*)(&pt[y+z]));
                                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~
src/encauth/ccm/ccm_memory.c:369:19: warning: cast from 'unsigned char *' to 'LTC_FAST_TYPE *' (aka 'unsigned long long *') increases required
      alignment from 1 to 8 [-Wcast-align]
                *((LTC_FAST_TYPE*)(&pt_real[y+z])) = *((LTC_FAST_TYPE*)(&pt[y+z])) & fastMask;
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/encauth/ccm/ccm_memory.c:369:56: warning: cast from 'unsigned char *' to 'LTC_FAST_TYPE *' (aka 'unsigned long long *') increases required
      alignment from 1 to 8 [-Wcast-align]
                *((LTC_FAST_TYPE*)(&pt_real[y+z])) = *((LTC_FAST_TYPE*)(&pt[y+z])) & fastMask;
                                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~

```

This is the cleanest trick I could think of to get it to work. Suggestions welcome.

karel-m commented:
I confirm that I am able to reproduce the above-mentioned warnings with clang and -Wcast-align (gcc with -Wcast-align does not warn about anything).

As for the solution the trick with `(whatever_long_type*)(void*)(..)` seems to be a de facto standard how to handle these clang warning.

:+1: from me

Please rebase this PR to the latest develop + fix also `testprof/store_test.c`

mattkelly commented:
Rebased onto develop, fixed testprof/store_test.c, and fixed some new occurrences of the warning in ccm_memory_ex.c and cbc_decrypt.c.

I think this is ready for merging.

karel-m commented:
looks good

@sjaeckel do you agree with merging this PR?

sjaeckel commented:
I pushed another commit to a branch pr/121 that suppresses some warnings generated by clang (at least on my version of clang). I'm not sure if that's the case for you as well, but if it is I'd propose to cherry-pick that commit e22a2c6193081ff9f23dba38d878e2664544536a directly to develop (as it's not part of the PR itself, but I pushed it to the branch out of laziness :) and because I created it when testing these changes... )
